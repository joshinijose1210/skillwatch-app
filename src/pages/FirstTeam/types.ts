export interface InputField {
    teamName: string;
    teamStatus: boolean;
    departmentRefId: number;
    teamRefId: number;
    error: string;
    organisationId: number;
    departmentId: number;
}

export interface StyledLottieProps {
    width?: string;
    paddingTop?: string;
    marginBottom?: string;
}
