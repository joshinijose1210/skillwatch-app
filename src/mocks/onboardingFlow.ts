export const onboardingFlowMockData = {
    firstDepartment: {
        url: 'https://dummy-api.example.com/api/departments/'
    },
    firstTeam: {
        url: 'https://dummy-api.example.com/api/teams/'
    },
    firstDesignation: {
        url: 'https://dummy-api.example.com/api/designation/'
    },
    firstEmployee: {
        url: 'https://dummy-api.example.com/api/onboarding-flow/1'
    },
    departments: {
        get: {
            url: 'https://dummy-api.example.com/api/departments/?organisationId=1&limit=10',
            someData: {
                totalDepartments: 2,
                departments: [
                    {
                        organisationId: 1,
                        id: 1,
                        departmentId: 'DEP12',
                        departmentName: 'Dep1',
                        departmentStatus: true,
                        departmentCreatedAt: 1693397266597
                    },
                    {
                        organisationId: 1,
                        id: 2,
                        departmentId: 'DEP11',
                        departmentName: 'Dep2',
                        departmentStatus: true,
                        departmentCreatedAt: 1692773271373
                    }
                ]
            }
        }
    },
    teams: {
        get: {
            url: 'https://dummy-api.example.com/api/teams/?organisationId=1&departmentId=1',
            responseCode: 200,
            data: {
                unlinkedTeamsCount: 0,
                totalTeams: 2,
                teams: [
                    {
                        organisationId: 1,
                        departmentId: 1,
                        departmentDisplayId: 'DEP3',
                        departmentName: 'Dep1',
                        departmentStatus: true,
                        id: 1,
                        teamId: 'T1',
                        teamName: 'Team1',
                        teamStatus: true,
                        teamCreatedAt: 1687865108632,
                        teamUpdatedAt: 1689765743375
                    },
                    {
                        organisationId: 1,
                        departmentId: 1,
                        departmentDisplayId: 'DEP3',
                        departmentName: 'Dep1',
                        departmentStatus: true,
                        id: 2,
                        teamId: 'T2',
                        teamName: 'Team2',
                        teamStatus: true,
                        teamCreatedAt: 1689711143838
                    }
                ]
            }
        }
    }
};
