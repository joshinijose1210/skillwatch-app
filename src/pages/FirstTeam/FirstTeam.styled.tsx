import { Button, Text } from '@medly-components/core';
import styled from 'styled-components';
import { StyledLottieProps } from './types';

export const LottieAndDivWrapper = styled.div`
    margin-top: 2rem;
    display: flex;
    align-items: center;
    column-gap: 10rem;
    @media (max-width: 900px) {
        display: flex;
        flex-direction: column;
    }
`;

export const StyledLottie = styled.div<StyledLottieProps>`
    width: ${({ width }) => width || '60%'};
    padding-top: ${({ paddingTop }) => paddingTop || '2rem'};
    margin-bottom: ${({ marginBottom }) => marginBottom || '0'};

    div {
        max-height: 70%;

        svg {
            display: inline-block;

            g {
                g:nth-child(78) {
                    opacity: 0;
                }
            }
        }
    }
`;

export const StyledAddFirstTeamDiv = styled.div`
    width: 100%;
    min-width: 440px;
    display: flex;
    align-items: flex-start;
`;

export const BorderDiv = styled.div`
    padding: 0.5rem;
    border: 2px solid ${({ theme }) => theme.colors.grey[200]};
    width: 100%;
`;

export const DivInsideDiv = styled.div`
    background-color: ${({ theme }) => theme.colors.grey[100]};
    padding: 2rem;
    display: flex;
    flex-direction: column;
    row-gap: 1.7rem;
`;

export const StyledText = styled(Text)``;

export const StyledButton = styled(Button)`
    width: fit-content;
    margin: 2rem 0;
`;

export const StyledSkipDiv = styled.div`
    display: flex;
    justify-content: end;
`;
export const StyledSkip = styled(Button)`
    width: fit-content;
    padding-right: 0;
`;

export const StyledGoBack = styled(Button)`
    width: fit-content;
    padding-left: 0;
`;

export const TextFieldDiv = styled.div`
    width: 50%;
`;

export const TeamAndDeleteButtonDiv = styled.div`
    display: flex;
    align-items: center;
`;

export const StyledAddTeam = styled(Text)<{ isDisabled?: boolean }>`
    color: ${({ theme, isDisabled }) => (isDisabled ? theme.colors.grey[400] : theme.colors.blue[500])};
    cursor: ${({ isDisabled }) => (isDisabled ? 'not-allowed' : 'pointer')};
    display: inline-block;
    :hover {
        text-decoration: ${({ isDisabled }) => (isDisabled ? 'none' : 'underline')};
    }
`;

export const AddTeamButtonDiv = styled.div``;

export const CancelAndSaveButtonDiv = styled.div`
    display: flex;
    justify-content: end;
`;

export const FlexDiv = styled.div`
    display: flex;
    column-gap: 1rem;
`;
