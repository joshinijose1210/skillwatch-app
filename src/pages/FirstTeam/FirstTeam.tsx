import { routeConstants } from '@constants';
import { Navigate } from 'react-router-dom';

export const FirstTeam = () => {
    // "Temporarily commented out as per requirement, might be added back later"

    // const {
    //     userDetails,
    //     inputFields,
    //     addTeam,
    //     View,
    //     handleAddAnotherTeam,
    //     handleAddTeam,
    //     handleCancelNo,
    //     handleCancelYes,
    //     handleInputChange,
    //     handleRemoveField,
    //     handleToggleClick,
    //     openCloseModal,
    //     handleSkip,
    //     modalState,
    //     handleSave,
    //     isAddTeamLoading,
    //     buttonDisabled,
    //     departmentList,
    //     handleDropdownChange,
    //     handleGoBack
    // } = useFirstTeam();
    // const { width } = useWindowSize();

    // if (userDetails && userDetails.onboardingFlow) {
    //     addToast({
    //         variant: 'success',
    //         header: 'Onboarding flow is completed'
    //     });
    return <Navigate to={routeConstants.root} />;
    // }

    // return (
    //     <PageContent>
    //         <ListHeader title={`Welcome, ${userDetails.firstName}`} />
    //         <LottieAndDivWrapper>
    //             <StyledAddFirstTeamDiv data-aos="fade-left">
    //                 <BorderDiv>
    //                     <DivInsideDiv>
    //                         <StyledText data-testid="title" textVariant="h4">
    //                             Org Admin <FromText textVariant="h4">&</FromText> Human Resources{' '}
    //                             <FromText textVariant="h4">
    //                                 teams are created for you. <br />
    //                                 Do you want to add more teams?
    //                             </FromText>
    //                         </StyledText>
    //                         {addTeam ? (
    //                             <>
    //                                 {inputFields &&
    //                                     inputFields.length > 0 &&
    //                                     inputFields.map((input: any, index: number) => {
    //                                         return (
    //                                             <TeamAndDeleteButtonDiv key={input.teamRefId}>
    //                                                 <TextFieldDesignation>
    //                                                     <StyledSingleSelect
    //                                                         options={departmentList}
    //                                                         variant="outlined"
    //                                                         size="M"
    //                                                         data-testid={`departmentName-${index}`}
    //                                                         placeholder="Select Department"
    //                                                         label="Department"
    //                                                         value={input.departmentId}
    //                                                         isSearchable
    //                                                         minWidth={width < 1450 ? '36%' : width < 1600 ? '40%' : '50%'}
    //                                                         onChange={val =>
    //                                                             val && handleDropdownChange(val, input.departmentRefId, input.teamName)
    //                                                         }
    //                                                     />
    //                                                     <TooltipDropdown
    //                                                         dataIds={[`department-input`]}
    //                                                         values={
    //                                                             input.departmentId && departmentList.length
    //                                                                 ? departmentList.filter(
    //                                                                       (item: OptionsType) => input.departmentId === item.value
    //                                                                   )
    //                                                                 : []
    //                                                         }
    //                                                     />
    //                                                     <StyledTextField
    //                                                         minWidth={width < 1450 ? '36%' : width < 1600 ? '40%' : '50%'}
    //                                                         name="teamName"
    //                                                         data-testid={`teamName-${index}`}
    //                                                         variant="outlined"
    //                                                         size="M"
    //                                                         label="Team Name"
    //                                                         placeholder="Enter Team Name"
    //                                                         type="text"
    //                                                         value={input.teamName}
    //                                                         onChange={(e: React.MouseEvent<HTMLInputElement>) =>
    //                                                             handleInputChange(e, input.teamRefId, input.departmentId)
    //                                                         }
    //                                                         errorText={input.error}
    //                                                     />

    //                                                     <StyledToggle
    //                                                         label="Active: "
    //                                                         labelVariant="h4"
    //                                                         data-testid={`activeToggle-${index}`}
    //                                                         labelWeight="Medium"
    //                                                         labelPosition="left"
    //                                                         onClick={(e: React.MouseEvent<HTMLInputElement>) =>
    //                                                             handleToggleClick(e, input.teamRefId)
    //                                                         }
    //                                                         checked={input.teamStatus}
    //                                                     />
    //                                                     <ClearButton
    //                                                         data-testid={`clearBtn-${index}`}
    //                                                         type="button"
    //                                                         onClick={() => handleRemoveField(input.teamRefId)}
    //                                                     >
    //                                                         <DeleteForeverIcon iconColor="red" />
    //                                                     </ClearButton>
    //                                                 </TextFieldDesignation>
    //                                             </TeamAndDeleteButtonDiv>
    //                                         );
    //                                     })}

    //                                 <AddTeamButtonDiv>
    //                                     <StyledAddTeam
    //                                         data-testid="addBtn"
    //                                         textVariant="h4"
    //                                         onClick={handleAddAnotherTeam}
    //                                         textAlign="center"
    //                                     >
    //                                         + Add Team
    //                                     </StyledAddTeam>
    //                                 </AddTeamButtonDiv>
    //                             </>
    //                         ) : (
    //                             <StyledButton data-testid="initAddBtn" variant="solid" onClick={handleAddTeam}>
    //                                 Add Team
    //                             </StyledButton>
    //                         )}
    //                         {addTeam ? (
    //                             <CancelAndSaveButtonDiv>
    //                                 <FlexDiv>
    //                                     <Button data-testid="cancelBtn" variant="outlined" onClick={openCloseModal}>
    //                                         Cancel
    //                                     </Button>
    //                                     <Button
    //                                         data-testid="saveBtn"
    //                                         variant="solid"
    //                                         disabled={buttonDisabled}
    //                                         onClick={handleSave}
    //                                         isLoading={isAddTeamLoading}
    //                                     >
    //                                         Save
    //                                     </Button>
    //                                 </FlexDiv>
    //                             </CancelAndSaveButtonDiv>
    //                         ) : (
    //                             <StyledSkipAndGoBackDiv>
    //                                 <StyledGoBack data-testid="go-back" variant="flat" onClick={handleGoBack}>
    //                                     &lt; Go Back
    //                                 </StyledGoBack>
    //                                 <StyledSkip data-testid="skiBtn" variant="flat" onClick={handleSkip}>
    //                                     Skip &gt;
    //                                 </StyledSkip>
    //                             </StyledSkipAndGoBackDiv>
    //                         )}
    //                     </DivInsideDiv>
    //                 </BorderDiv>
    //             </StyledAddFirstTeamDiv>
    //             <StyledLottie paddingTop="1rem" width="65%" marginBottom="8rem" data-aos="fade-right">
    //                 {View}
    //             </StyledLottie>
    //         </LottieAndDivWrapper>
    //         <Modal open={modalState} onCloseModal={openCloseModal}>
    //             <StyledModalHeader>
    //                 <Text textVariant="h2">Confirm</Text>
    //             </StyledModalHeader>
    //             <StyledModalContent>
    //                 <Text textVariant="body1" textWeight="Medium">
    //                     Are you sure you wish to cancel? Any unsaved changes will be lost.
    //                 </Text>
    //             </StyledModalContent>
    //             <StyledModalActions>
    //                 <CancelButton data-testid="modal-yes" variant="outlined" onClick={handleCancelYes}>
    //                     Yes
    //                 </CancelButton>
    //                 <Button data-testid="modal-no" onClick={handleCancelNo} variant="outlined">
    //                     No
    //                 </Button>
    //             </StyledModalActions>
    //         </Modal>
    //     </PageContent>
    // );
};
