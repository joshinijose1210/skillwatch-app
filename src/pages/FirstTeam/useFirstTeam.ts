import { Department } from '@common';
import { routeConstants } from '@constants';
import { addToast } from '@medly-components/core';
import { Option } from '@medly-components/core/dist/es/components/MultiSelect/types';
import { useAppSelector } from '@slice';
import { useGetDepartmentQuery, usePostTeamMutation } from '@slice/services';
import { performValidation } from '@utils/validations';
import { useLottie } from 'lottie-react';
import { useCallback, useEffect, useMemo, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import addTeamAnimation from '../../constants/images/animations/addTeamAnimation.json';
import { InputField } from './types';

export const useFirstTeam = () => {
    const userDetails = useAppSelector(state => state.user);
    const [addTeam, setAddTeam] = useState(false);
    const [inputFields, setInputFields] = useState<InputField[]>([
        {
            teamName: '',
            teamStatus: true,
            departmentRefId: new Date().getTime(),
            teamRefId: new Date().getTime(),
            error: '',
            organisationId: userDetails.organisationId,
            departmentId: -1
        }
    ]);
    const [modalState, setModalState] = useState(false);
    const [addFirstTeam, { isSuccess: isAddSuccess, data: responseData, isLoading: isAddTeamLoading }] = usePostTeamMutation();

    const { data: department, isSuccess: isDepartmentReceived } = useGetDepartmentQuery(
        {
            path: '',
            params: [
                { name: 'organisationId', value: userDetails.organisationId },
                { name: 'limit', value: 10 }
            ]
        },
        { refetchOnMountOrArgChange: true }
    );

    const departmentList = useMemo(
        () =>
            isDepartmentReceived
                ? department?.departments
                      .filter((department: Department) => department.departmentStatus === true)
                      .map((department: Department) => {
                          return { value: department.id, label: department.departmentName };
                      })
                      .sort((a: Option, b: Option) => a.label.localeCompare(b.label))
                : [],
        [isDepartmentReceived, department?.departments]
    );

    const navigate = useNavigate();
    const defaultOptions = {
        loop: true,
        autoplay: true,
        animationData: addTeamAnimation,
        rendererSettings: {
            preserveAspectRatio: 'xMidYMid slice'
        }
    };

    const { View } = useLottie(defaultOptions);

    const handleAddTeam = () => {
        setAddTeam(true);
    };

    const buttonDisabled = useMemo(() => {
        let disabled = false;
        inputFields.forEach((element: InputField) => {
            if ((element.error && element.error.length > 0) || element.teamName.length < 1) {
                disabled = true;
            }
        });
        return disabled;
    }, [inputFields]);

    const handleInputChange = (e: React.MouseEvent<HTMLInputElement>, id: number, departmentId: number) => {
        e.preventDefault();
        const { name, value } = e.currentTarget;
        const newArray = inputFields.map((team: InputField) => {
            if (team.teamRefId === id) {
                const existingTeamName = inputFields.find(
                    (field: InputField) => field.teamRefId !== id && field.teamName === value && field.departmentId === departmentId
                );
                const err = performValidation(name, value);

                return {
                    ...team,
                    teamName: value,
                    error: existingTeamName ? 'Team name already entered for the department' : err ? err : ''
                };
            } else return team;
        });
        setInputFields(newArray);
    };

    const handleDropdownChange = (departmentId: number, id: number, team: string) => {
        const newArray = inputFields.map((department: InputField) => {
            if (department.departmentRefId === id) {
                const existingDepartmentName = inputFields.find(
                    (field: InputField) => field.teamRefId !== id && field.teamName === team && field.departmentId === departmentId
                );

                return {
                    ...department,
                    departmentId,
                    error: existingDepartmentName ? 'Team name already entered for the Department.' : ''
                };
            } else return department;
        });
        setInputFields(newArray);
    };

    const handleAddAnotherTeam = () => {
        setInputFields([
            ...inputFields,
            {
                teamName: '',
                teamStatus: true,
                teamRefId: new Date().getTime(),
                departmentRefId: new Date().getTime(),
                error: '',
                organisationId: userDetails.organisationId,
                departmentId: -1
            }
        ]);
    };

    const handleRemoveField = (removeId: number) => {
        const newInputFields = inputFields.filter(({ teamRefId }: InputField) => {
            return teamRefId !== removeId;
        });
        setInputFields(newInputFields);
    };

    useEffect(() => {
        if (inputFields.length === 0) {
            setAddTeam(false);
            setInputFields([
                {
                    teamName: '',
                    teamStatus: true,
                    departmentRefId: new Date().getTime(),
                    teamRefId: new Date().getTime(),
                    error: '',
                    organisationId: userDetails.organisationId,
                    departmentId: -1
                }
            ]);
        }
    }, [inputFields.length, userDetails.organisationId]);

    const handleSave = () => {
        addFirstTeam({ actionBy: userDetails.id, ipAddress: '216.24.57.253:443', teams: inputFields });
    };

    const handleToggleClick = (e: React.MouseEvent<HTMLInputElement>, id: number) => {
        const newArray = inputFields.map((team: InputField) => {
            if (team.teamRefId === id) {
                return {
                    ...team,
                    teamStatus: e.currentTarget.checked
                };
            } else return team;
        });
        setInputFields(newArray);
    };

    const handleSkip = useCallback(() => {
        navigate(routeConstants.firstDesignation);
    }, [navigate]);

    const handleGoBack = () => {
        navigate(routeConstants.firstDepartment);
    };

    const openCloseModal = () => {
        if (inputFields && inputFields[0] && inputFields[0].teamName.length > 0) {
            setModalState(!modalState);
        } else setInputFields([]);
    };

    const handleCancelYes = () => {
        setInputFields([]);
        openCloseModal();
    };

    const handleCancelNo = () => {
        openCloseModal();
    };

    useEffect(() => {
        if (responseData) {
            if (responseData.existingTeam) {
                setInputFields((prevFields: any) =>
                    prevFields.map((field: any) => ({
                        ...field,
                        error: responseData.existingTeam.includes(field.teamName) ? `Team ${field.teamName} already exists` : field.error
                    }))
                );
            }
            if (responseData.addedTeam) {
                addToast({
                    variant: 'success',
                    header: `Team ${responseData.addedTeam} added successfully`,
                    timer: 3000
                });
                setInputFields((prevFields: any) => prevFields.filter((field: any) => !responseData.addedTeam.includes(field.teamName)));
                isAddSuccess && !responseData.existingTeam && handleSkip();
            }
        }
    }, [handleSkip, isAddSuccess, responseData]);

    return {
        userDetails,
        inputFields,
        addTeam,
        View,
        handleAddAnotherTeam,
        handleAddTeam,
        handleCancelNo,
        handleCancelYes,
        handleInputChange,
        handleRemoveField,
        handleToggleClick,
        openCloseModal,
        handleSkip,
        modalState,
        handleSave,
        isAddTeamLoading,
        buttonDisabled,
        departmentList,
        handleDropdownChange,
        handleGoBack
    };
};
