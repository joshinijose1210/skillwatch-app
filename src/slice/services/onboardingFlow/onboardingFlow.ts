import apiUrls from '@constants/apiUrls';
import { createApi } from '@reduxjs/toolkit/query/react';
import { axiosBaseQuery } from '@utils/axiosBaseQuery';

export const onboardingFlowApi = createApi({
    reducerPath: 'onboardingFlowApi',
    baseQuery: axiosBaseQuery({
        baseUrl: apiUrls.onboardingFlow
    }),
    tagTypes: ['onboardingFlow'],

    endpoints: builder => ({
        postOnboardingFlow: builder.mutation({
            query: payload => ({
                url: `/${payload.path}`,
                method: 'PUT',
                data: payload.data,
                headers: {
                    'Content-type': 'application/json; charset=UTF-8'
                }
            })
        })
    })
});

export const { usePostOnboardingFlowMutation } = onboardingFlowApi;
