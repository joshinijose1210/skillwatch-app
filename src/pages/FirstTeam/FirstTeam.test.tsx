import { RTKRender, cleanup, fireEvent, screen, waitFor } from '@utils/test-utils';
import { FirstTeam } from './FirstTeam';
import MockAdapter from 'axios-mock-adapter';
import { apiInstance } from '@utils';
import { orgAdminMockState } from '../../mocks/userMockState';
import { onboardingFlowMockData } from '@mocks/onboardingFlow';
const mock = new MockAdapter(apiInstance, { onNoMatch: 'throwException' });
const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
    Navigate: jest.fn(() => <div data-testid="mock-navigate">Navigate component</div>)
}));

beforeAll(() => {
    const { url, someData } = onboardingFlowMockData.departments.get;
    mock.onGet(url).replyOnce(200, someData);
});

beforeEach(mock.reset);

afterEach(cleanup);

describe('FirstTeam Component', () => {
    it('should take snapshot', () => {
        const { container } = RTKRender(<FirstTeam />, { initialState: { ...orgAdminMockState, onboardingFlow: false } });
        expect(container).toMatchSnapshot();
    });

    it('should render properly', () => {
        RTKRender(<FirstTeam />, { initialState: { ...orgAdminMockState, onboardingFlow: false } });
        expect(screen.getByText('teams are created for you. Do you want to add more teams?')).toBeInTheDocument();
        expect(screen.getByTestId('initAddBtn')).toBeEnabled();
    });

    it('should navigate when skip is clicked', () => {
        RTKRender(<FirstTeam />, { initialState: { ...orgAdminMockState, onboardingFlow: false } });
        const skipBtn = screen.getByTestId('skiBtn');
        expect(skipBtn).toBeEnabled();
        fireEvent.click(skipBtn);
        expect(mockNavigate).toBeCalledWith('/set-up-designation');
    });

    it('should navigate when go-back is clicked', () => {
        RTKRender(<FirstTeam />, { initialState: { ...orgAdminMockState, onboardingFlow: false } });
        const goBackBtn = screen.getByTestId('go-back');
        expect(goBackBtn).toBeEnabled();
        fireEvent.click(goBackBtn);
        expect(mockNavigate).toBeCalledWith('/set-up-department');
    });

    it('should all function work properly and add department', async () => {
        mock.onPost(onboardingFlowMockData.firstTeam.url).replyOnce(200, {
            addedTeam: ['Test team']
        });
        const { url, someData } = onboardingFlowMockData.departments.get;
        mock.onGet(url).replyOnce(200, someData);
        RTKRender(<FirstTeam />, { initialState: { ...orgAdminMockState, onboardingFlow: false } });
        fireEvent.click(screen.getByTestId('initAddBtn'));
        await screen.findByTestId('clearBtn-0');
        fireEvent.click(screen.getByTestId('clearBtn-0'));
        await waitFor(() => expect(screen.queryByTestId('clearBtn-0')).not.toBeInTheDocument());
        fireEvent.click(screen.getByTestId('initAddBtn'));
        await screen.findByTestId('clearBtn-0');
        fireEvent.click(screen.getByTestId('cancelBtn'));
        await waitFor(() => expect(screen.queryByTestId('clearBtn-0')).not.toBeInTheDocument());
        fireEvent.click(screen.getByTestId('initAddBtn'));
        await screen.findByTestId('clearBtn-0');
        fireEvent.click(screen.getByTestId('departmentName-0'));
        await screen.findByText('Dep1');
        fireEvent.click(screen.getByText('Dep1'));
        fireEvent.change(screen.getByTestId('teamName-0'), { target: { value: 'Test team' } });
        fireEvent.click(screen.getByTestId('activeToggle-0'));
        fireEvent.click(screen.getByTestId('activeToggle-0'));
        fireEvent.click(screen.getByTestId('saveBtn'));
        expect((await screen.findAllByText('Team Test team added successfully'))[0]).toBeInTheDocument();
    });

    it('should modal work propely', async () => {
        const { url, someData } = onboardingFlowMockData.departments.get;
        mock.onGet(url).replyOnce(200, someData);
        RTKRender(<FirstTeam />, { initialState: { ...orgAdminMockState, onboardingFlow: false } });
        fireEvent.click(screen.getByTestId('initAddBtn'));
        await screen.findByTestId('clearBtn-0');
        fireEvent.click(screen.getByTestId('departmentName-0'));
        await screen.findByText('Dep1');
        fireEvent.click(screen.getByText('Dep1'));
        const departmentInput = screen.getByTestId('teamName-0');
        fireEvent.change(departmentInput, { target: { value: 'Test team' } });
        fireEvent.click(screen.getByTestId('cancelBtn'));
        await screen.findByTestId('modal-no');
        fireEvent.click(screen.getByTestId('modal-no'));
        await waitFor(() => expect(document.body).toHaveStyle({ overflow: 'unset' }));
        expect(departmentInput).toHaveValue('Test team');
        fireEvent.click(screen.getByTestId('cancelBtn'));
        await screen.findByTestId('modal-yes');
        fireEvent.click(screen.getByTestId('modal-yes'));
        await waitFor(() => expect(document.body).toHaveStyle({ overflow: 'unset' }));
        expect(screen.queryByTestId('clearBtn-0')).not.toBeInTheDocument();
    });

    it('should adding multiple department work propely', async () => {
        RTKRender(<FirstTeam />, { initialState: { ...orgAdminMockState, onboardingFlow: false } });
        fireEvent.click(screen.getByTestId('initAddBtn'));
        expect(await screen.findByTestId('clearBtn-0')).toBeInTheDocument();
        fireEvent.click(screen.getByTestId('addBtn'));
        expect(await screen.findByTestId('clearBtn-1')).toBeInTheDocument();
    });

    it('should show an error when department already exist', async () => {
        mock.onPost(onboardingFlowMockData.firstTeam.url).replyOnce(200, {
            existingTeam: ['Test team']
        });
        const { url, someData } = onboardingFlowMockData.departments.get;
        mock.onGet(url).replyOnce(200, someData);
        RTKRender(<FirstTeam />, { initialState: { ...orgAdminMockState, onboardingFlow: false } });
        fireEvent.click(screen.getByTestId('initAddBtn'));
        await screen.findByTestId('clearBtn-0');
        fireEvent.click(screen.getByTestId('departmentName-0'));
        await screen.findByText('Dep1');
        fireEvent.click(screen.getByText('Dep1'));
        fireEvent.change(screen.getByTestId('teamName-0'), { target: { value: 'Test team' } });
        fireEvent.click(screen.getByTestId('saveBtn'));
        expect(await screen.findByText('Team Test team already exists')).toBeInTheDocument();
    });

    it('should render root component if onboarding flow is false', async () => {
        RTKRender(<FirstTeam />, { initialState: { ...orgAdminMockState, onboardingFlow: true } });
        expect(screen.getAllByText('Onboarding flow is completed')[0]).toBeInTheDocument();
    });

    it('should add different teams at once', async () => {
        mock.onPost(onboardingFlowMockData.firstTeam.url).replyOnce(200, {
            existingTeam: ['Test team'],
            addedTeam: ['Test new', 'Team team1']
        });
        const { url, someData } = onboardingFlowMockData.departments.get;
        mock.onGet(url).replyOnce(200, someData);
        RTKRender(<FirstTeam />, { initialState: { ...orgAdminMockState, onboardingFlow: false } });
        fireEvent.click(screen.getByTestId('initAddBtn'));
        await screen.findByTestId('clearBtn-0');
        fireEvent.click(screen.getByTestId('departmentName-0'));
        await screen.findByText('Dep1');
        fireEvent.click(screen.getByText('Dep1'));
        fireEvent.change(screen.getByTestId('teamName-0'), { target: { value: 'Test team' } });
        fireEvent.change(screen.getByTestId('teamName-0'), { target: { value: 'Test team1' } });
        fireEvent.click(screen.getByTestId('addBtn'));
        fireEvent.change(screen.getByTestId('teamName-1'), { target: { value: 'Test new' } });
        fireEvent.click(screen.getByTestId('saveBtn'));

        expect(await screen.findByText('Team Test new,Team team1 added successfully')).toBeInTheDocument();
    });

    it('should throw error when adding a team to a department that already is being created in that department', async () => {
        const { url, someData } = onboardingFlowMockData.departments.get;
        mock.onGet(url).replyOnce(200, someData);
        RTKRender(<FirstTeam />, { initialState: { ...orgAdminMockState, onboardingFlow: false } });
        fireEvent.click(screen.getByTestId('initAddBtn'));
        await screen.findByTestId('clearBtn-0');
        fireEvent.click(screen.getByTestId('departmentName-0'));
        await screen.findByText('Dep1');
        fireEvent.click(screen.getByText('Dep1'));
        fireEvent.change(screen.getByTestId('teamName-0'), { target: { value: 'Test team' } });
        fireEvent.click(screen.getByTestId('addBtn'));
        fireEvent.click(screen.getByTestId('departmentName-1'));
        fireEvent.click(screen.getByText('Dep1'));
        fireEvent.change(screen.getByTestId('teamName-1'), { target: { value: 'Test team' } });

        expect(await screen.findByText('Team name already entered for the department')).toBeInTheDocument();
    });

    it('should add different teams at once being inactive', async () => {
        mock.onPost(onboardingFlowMockData.firstTeam.url).replyOnce(200, {
            existingTeam: ['Test team'],
            addedTeam: ['Test new', 'Team team1']
        });
        const { url, someData } = onboardingFlowMockData.departments.get;
        mock.onGet(url).replyOnce(200, someData);
        RTKRender(<FirstTeam />, { initialState: { ...orgAdminMockState, onboardingFlow: false } });
        fireEvent.click(screen.getByTestId('initAddBtn'));
        await screen.findByTestId('clearBtn-0');
        fireEvent.click(screen.getByTestId('departmentName-0'));
        await screen.findByText('Dep1');
        fireEvent.click(screen.getByText('Dep1'));
        fireEvent.change(screen.getByTestId('teamName-0'), { target: { value: 'Test team' } });
        fireEvent.change(screen.getByTestId('teamName-0'), { target: { value: 'Test team1' } });
        fireEvent.click(screen.getByTestId('addBtn'));
        fireEvent.change(screen.getByTestId('teamName-1'), { target: { value: 'Test new' } });
        fireEvent.click(screen.getByTestId('saveBtn'));
        fireEvent.click(screen.getByTestId('activeToggle-0'));
        fireEvent.click(screen.getByTestId('activeToggle-1'));

        expect(await screen.findByText('Team Test new,Team team1 added successfully')).toBeInTheDocument();
    });

    it('should throw error when adding a wrong team name', async () => {
        const { url, someData } = onboardingFlowMockData.departments.get;
        mock.onGet(url).replyOnce(200, someData);
        RTKRender(<FirstTeam />, { initialState: { ...orgAdminMockState, onboardingFlow: false } });
        fireEvent.click(screen.getByTestId('initAddBtn'));
        await screen.findByTestId('clearBtn-0');
        fireEvent.click(screen.getByTestId('departmentName-0'));
        await screen.findByText('Dep1');
        fireEvent.click(screen.getByText('Dep1'));
        fireEvent.change(screen.getByTestId('teamName-0'), { target: { value: 'Test!@#$%' } });
        fireEvent.click(screen.getByTestId('saveBtn'));
        expect(await screen.findByText(/Special characters are not allowed./i)).toBeInTheDocument();
    });

    it('should add different teams at once to different departments', async () => {
        mock.onPost(onboardingFlowMockData.firstTeam.url).replyOnce(200, {
            existingTeam: ['Test team'],
            addedTeam: ['Test new', 'Team team1']
        });
        const { url, someData } = onboardingFlowMockData.departments.get;
        mock.onGet(url).replyOnce(200, someData);
        RTKRender(<FirstTeam />, { initialState: { ...orgAdminMockState, onboardingFlow: false } });
        fireEvent.click(screen.getByTestId('initAddBtn'));
        await screen.findByTestId('clearBtn-0');
        fireEvent.click(screen.getByTestId('departmentName-0'));
        await screen.findByText('Dep1');
        fireEvent.click(screen.getByText('Dep1'));
        fireEvent.change(screen.getByTestId('teamName-0'), { target: { value: 'Test team1' } });
        fireEvent.click(screen.getByTestId('addBtn'));
        fireEvent.click(screen.getByTestId('departmentName-1'));
        fireEvent.click(screen.getByText('Dep2'));

        fireEvent.change(screen.getByTestId('teamName-1'), { target: { value: 'Test new' } });
        fireEvent.click(screen.getByTestId('saveBtn'));
        expect((await screen.findAllByText('Team Test new,Team team1 added successfully'))[0]).toBeInTheDocument();
    });
});
