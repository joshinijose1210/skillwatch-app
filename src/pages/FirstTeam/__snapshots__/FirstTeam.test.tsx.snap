// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirstTeam Component should take snapshot 1`] = `
<div>
  <div
    data-testid="mock-navigate"
  >
    Navigate component
  </div>
</div>
`;
